src/shared/ui/common/region.ts

```diff
@@ -0,0 +1,56 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as nls from 'vscode-nls'
+import { ExtContext } from '../../extensions'
+import { Region } from '../../regions/endpoints'
+import { getRegionsForActiveCredentials } from '../../regions/regionUtilities'
+import { PrompterButtons } from '../buttons'
+import { createQuickPick, QuickPickPrompter } from '../pickerPrompter'
+
+const localize = nls.loadMessageBundle()
+
+type RegionContext = Pick<ExtContext, 'regionProvider' | 'awsContext'>
+type RegionPrompterOptions = {
+    currentRegion?: Region | string
+    title?: string
+    buttons?: PrompterButtons<Region>
+    /** Filter regions based off service availability */
+    serviceId?: string
+}
+
+export function createRegionPrompter(regions: Region[], options: RegionPrompterOptions): QuickPickPrompter<Region>
+export function createRegionPrompter(context: RegionContext, options: RegionPrompterOptions): QuickPickPrompter<Region>
+export function createRegionPrompter(
+    regions: Region[] | RegionContext,
+    options: RegionPrompterOptions = {}
+): QuickPickPrompter<Region> {
+    if (!Array.isArray(regions)) {
+        const { awsContext, regionProvider } = regions
+        regions = getRegionsForActiveCredentials(awsContext, regionProvider)
+        options.currentRegion = options.currentRegion ?? awsContext.getCredentialDefaultRegion()
+
+        if (options.serviceId !== undefined) {
+            regions = regions.filter(r => regionProvider.isServiceInRegion(options.serviceId!, r.id))
+        }
```

Comment:
This feels super gross with the fact that you're reassigning `regions` to an entirely different type than it starts as. Is it worth just creating a `let` to hold the array of regions and assign the value of `regions` to the `let` var in an else statement?

Reply:
I'm just removing the array option, doesn't serve much of a purpose.